### PR TITLE
Add binary_frame hook

### PR DESCRIPTION
**Pull request by aminmarashi**

### CODE REVIEW

**Comment by jy-binary** [comment]
failing tests

File: lib/Mojo/WebSocketProxy/Dispatcher.pm [hunk 1]
```diff
@@ -61,7 +61,16 @@ sub open_connection {
 
     $config->{opened_connection}->($c) if $config->{opened_connection};
 
-    $c->on(json => \&on_message);
+    $c->on(json => sub {
+        my ($d, $hash) = @_;
```
Comment (basiliscos) [inline lib/Mojo/WebSocketProxy/Dispatcher.pm]: what is `$d` and `$hash` of what?

File: lib/Mojo/WebSocketProxy/Dispatcher.pm [hunk 2]
```diff
@@ -61,7 +61,16 @@ sub open_connection {
...
+        on_message(@_) if $hash;
```
Comment (basiliscos) [inline lib/Mojo/WebSocketProxy/Dispatcher.pm]: what this condition means? What if it does not match? We silently ignore the result?

File: lib/Mojo/WebSocketProxy/Dispatcher.pm [hunk 3]
```diff
@@ -61,7 +61,16 @@ sub open_connection {
...
+    });
+
+    $c->on(binary => sub {
+        my ($d, $bytes) = @_;
+        $config->{binary_frame}(@_) if $bytes;
```
Comment (basiliscos) [inline lib/Mojo/WebSocketProxy/Dispatcher.pm]: consider adding tests for that case + update `Changes`

**Comment by basiliscos** [comment]
@jy-binary @aminmarashi tests fail because of issue with 3rd party library. @aminmarashi , please execute tests locally (`dzil test`) and paste results here to be sure they pass.

**Comment by aminmarashi** [comment]
This is the local test result.

```
[DZ] building distribution under .build/rrUpCyYQrc for installation
[DZ] beginning to build Mojo-WebSocketProxy
[DZ] guessing dist's main_module is lib/Mojo/WebSocketProxy.pm
[VersionFromModule] dist version 0.06 (from lib/Mojo/WebSocketProxy.pm)
[OurPkgVersion] Skipping: "lib/Mojo/WebSocketProxy.pm" has no "# VERSION" comment
[Prereqs::FromCPANfile] Parsing 'cpanfile' to extract prereqs
[GitHub::Meta] Getting GitHub repository info
[DZ] writing Mojo-WebSocketProxy in .build/rrUpCyYQrc
Checking if your kit is complete...
Looks good
Generating a Unix-style Makefile
Writing Makefile for Mojo::WebSocketProxy
Writing MYMETA.yml and MYMETA.json
cp lib/Mojo/WebSocketProxy/CallingEngine.pm blib/lib/Mojo/WebSocketProxy/CallingEngine.pm
cp lib/Mojo/WebSocketProxy/Dispatcher.pm blib/lib/Mojo/WebSocketProxy/Dispatcher.pm
cp lib/Mojolicious/Plugin/WebSocketProxy.pm blib/lib/Mojolicious/Plugin/WebSocketProxy.pm
cp lib/Mojo/WebSocketProxy/Parser.pm blib/lib/Mojo/WebSocketProxy/Parser.pm
cp lib/Mojo/WebSocketProxy/Config.pm blib/lib/Mojo/WebSocketProxy/Config.pm
cp lib/Mojo/WebSocketProxy.pm blib/lib/Mojo/WebSocketProxy.pm
PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/00-check-deps.t ................. ok
t/00-compile.t .................... ok
t/00-report-prereqs.t ............. ok
t/05_basic.t ...................... ok
t/10-before_forward.t ............. ok
t/11-timeout.t .................... ok
t/12-before_send_api_response.t ... ok
t/13-action-not-found.t ........... ok
t/14-bad-request.t ................ ok
t/15-simple-success.t ............. ok
t/16-auth-error-sample.t .......... ok
t/17-instead_of_forward.t ......... ok
t/18-per-action_before-forward.t .. ok
t/19-rpc_response_cb.t ............ ok
t/20-after_got_rpc_response.t ..... ok
t/21-mojo-stash.t ................. ok
t/22-override-response.t .......... ok
t/23-binary-frame.t ............... ok
All tests successful.
Files=18, Tests=147, 17 wallclock secs ( 0.07 usr  0.04 sys +  7.02 cusr  0.74 csys =  7.87 CPU)
Result: PASS
[DZ] all's well; removing .build/rrUpCyYQrc```

**Comment by basiliscos** [comment]
@aminmarashi Thanks. That's nice. Could. you, please, add some real binary file for testing (i.e. take `binary logo.jpg`) - that's will make tests bullet-proof.

**Comment by aminmarashi** [comment]
@basiliscos I changed the test accordingly.

**Comment by basiliscos** [comment]
@aminmarashi I see you put the numbers. I'd suggest to to try on real binary file. You can load it easily in perl tests as :

- put it somewhere in `t/data/logo.png`

- then

```perl
my $bytes = path('t/data/logo.png')->slurp;
# ... pack data into frame
my $received_data = ... ;
is $bytes, $received_data;
```

**Comment by aminmarashi** [comment]
@basiliscos I hope you like it :)

**Comment by basiliscos** [comment]
@aminmarashi Perfect!

**Comment by jy-binary** [comment]
@aminmarashi please check test failures

**Comment by jy-binary** [comment]
@aminmarashi can you check why tests fail?

**Comment by aminmarashi** [comment]
@jy-binary Not sure, but @basiliscos mentioned it's because of a third-party problem. The local tests pass though.

**Comment by basiliscos** [comment]
Yes, test fail on master, because of abandoned 3rd-party library, which does not supports `Mojolicious-v7`. We have local fix in our private cpan, though.

Fails on master (for previous public release):

http://matrix.cpantesters.org/?dist=Mojo-WebSocketProxy+0.06

**Pull request merged**